src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -419,6 +432,35 @@ static bool simplify_log(shared_ptr<Node> n)
     return false;
 }
 
+// optimizes `gather->shapeof` into `shapeof->gather` for 0D gather indices
+static bool simplify_gather_shapeof(shared_ptr<Node> node)
+{
+    auto shapeof = as_type_ptr<op::v0::ShapeOf>(node);
+    auto gather = as_type_ptr<op::v1::Gather>(shapeof->input_value(0).get_node_shared_ptr());
+    if (!gather)
+    {
+        return false;
+    }
+    auto indices = gather->input_value(1).get_node_shared_ptr();
+    auto gather_in_ps = gather->get_input_partial_shape(0);
+    auto indices_ps = indices->get_output_partial_shape(0);
+    auto axis = gather->get_axis();
+    if (gather_in_ps.rank().is_dynamic() || indices_ps.rank().is_dynamic() ||
+        indices_ps.rank().get_length() != 0 || axis == op::v1::Gather::AXIS_NOT_SET_VALUE)
```

Comment:
shouldn't we transform `Gather->ShapeOf` to `Concat(ShapeOf(data)[:axis], ShapeOf(indices) + ShapeOf(data)[axis + 1:]`? and as an exceptional case leave only `indices.shape`? does it sounds sane?

Reply:
this would be `Concat(Gather(ShapeOf(data),[:axis]), ShapeOf(indices) + Gather(ShapeOf(data),[axis + 1:])` right ? resulting in several ops. If this is needed, we could do it.